tests/passing/expectExtend-illegalWrites.spec.lua

```diff
@@ -0,0 +1,68 @@
+-- luacheck: globals describe beforeAll expect it
+
+local noOptMatcher = function(_received, _expected)
+	return {
+		message = "",
+		pass = true,
+	}
+end
+
+local ERROR_CANNOT_OVERWRITE = "Cannot overwrite matcher"
+local ERROR_CANNOT_START_WITH = "Matchers cannot start with"
+
+local runTest = function(expectedError)
+	return function()
+		local success, message = pcall(function()
+			expect()
+		end)
+
+		assert(success == false, "should have been thrown")
+		assert(message:match(expectedError), string.format("\nUnexpected error:\n%s", message))
+	end
+end
+
+return function()
+	describe("attempt to overwrite default", function()
+		beforeAll(function()
+			expect.extend({
+				-- This should throw since `ok` is a default matcher
+				ok = noOptMatcher,
+			})
+		end)
+
+		it("SHOULD fail with expected error message", runTest(ERROR_CANNOT_OVERWRITE))
```

Comment:
I'm kind of surprised this works. It looks like it should fail in the beforeAll where extend is called instead of in the pcall where expect is called.

Reply:
I agree with your comments.


The reason this works currently is because when we call `expect.extend`, we're only calling the extend from the ExpectationContext.lua file, not the Expectation.lua file.

ExpectationContext.lua file does not have access to private members of Expectation.lua, so we arent able to check for default fields easily.

Since we only create a Expectation object when we invoke `expect()` we aren't able to check against this class until then.

Ideally, we'd throw much earlier in beforeAll, but I was wondering if you had any advice or ideas on how we can do that in a clean approachable way from ExpectationContext.

Reply:
Updated, no longer the case 👍